### PR TITLE
Introduce RelResult rather than Vec<Vec<TypedValue>>.

### PR DESCRIPTION
**Pull request by rnewman**

### CODE REVIEW

**Comment by rnewman** [comment]
I cleaned up and improved this, @fluffyemily .

File: core/src/types.rs [hunk 1]
```diff
@@ -172,6 +186,134 @@ pub enum TypedValue {
     Uuid(Uuid),                        // It's only 128 bits, so this should be acceptable to clone.
 }
 
+/// The values stored in a `StructuredMap` can be:
```
Comment (ncalexan) [inline core/src/types.rs]: The comment refers to `StructuredMap`, but it's attached to `Binding`.  Huh?

File: core/src/types.rs [hunk 2]
```diff
@@ -172,6 +186,134 @@ pub enum TypedValue {
...
+/// * Vecs of structured values, for multi-valued component attributes or nested expressions.
+/// * Single structured values, for single-valued component attributes or nested expressions.
+/// * Single typed values, for simple attributes.
+///   Note that the `TypedValue` enum is inlined here for convenience: `Binding` is
+///   an expansion of it.
+///
+/// Datomic also supports structured _bindings_; at present Mentat does not, but this type
+/// would also serve that purpose.
+///
+/// Note that maps are not ordered….
```
Comment (ncalexan) [inline core/src/types.rs]: It seems like you think this is significant.  If it is, explain why.

File: core/src/types.rs (hunk 1 above)
Comment (rnewman) [reply core/src/types.rs]: This comment is cleaned up later.

File: core/src/types.rs (hunk 2 above)
Comment (rnewman) [reply core/src/types.rs]: … 'cos it means this doesn't implement `Ord`.

**Comment by rnewman** [comment]
Using this PR for the `RelResult` change, then landing the file split as a separate commit, then I'll open and land another PR for `Binding`.

**Pull request merged**